src/java.base/share/classes/jdk/internal/vm/annotation/Stable.java

```diff
@@ -28,56 +28,191 @@
  * More specifically, the HotSpot VM will process non-null stable fields (final
  * or otherwise) in a similar manner to static final fields with respect to
  * promoting the field's value to a constant.  Thus, placing aside the
  * differences for null/non-null values and arrays, a final stable field is
  * treated as if it is really final from both the Java language and the HotSpot
  * VM.
  * <p>
- * It is (currently) undefined what happens if a field annotated as stable
- * is given a third value (by explicitly updating a stable field, a component of
- * a stable array, or a final stable field via reflection or other means).
- * Since the HotSpot VM promotes a non-null component value to constant, it may
- * be that the Java memory model would appear to be broken, if such a constant
- * (the second value of the field) is used as the value of the field even after
- * the field value has changed (to a third value).
+ * After constant folding, the compiler can make use of may aspects of
+ * the object: Its dynamic type, its length (if it is an array), and
+ * the values of its fields (if they are themselves constants, either
+ * final or stable).  It is in general a bad idea to reset such
+ * variables to any other value, since compiled code might have folded
+ * an earlier stored value, and will never detect the reset value.
+ * <p>
+ * The HotSpot interpreter is not fully aware of stable annotations,
+ * and treats annotated fields (and any affected arrays) as regular
+ * mutable variables.  Thus, a field annotated as {@code @Stable} may
+ * be given a series of values, by explicit assignment, by reflection,
+ * or by some other means.  If the HotSpot compiler observes one of
+ * these values, and constant-folds it in the setting of some
+ * particular compilation task, then in some contexts (execution of
+ * fully optimized code) the field will appear to have one
+ * "historical" value, while in others (less optimized contexts) the
+ * field will appear to have a more recent value.  (And it is no good
+ * to try to "reset" a stable value by storing its default again,
+ * because there is currently no way to find and deoptimize any and
+ * all affected compiled code.)  Race conditions would make this even
+ * more complex, since with races there is no definable "most recent"
+ * value.
+ * <p>
+ * Note also each compiliation task makes its own decisions about
+ * whether to observe stable variable values, and how aggressively to
+ * constant-fold them.  And a method that uses a stable variable might
+ * be inlined by many different compilation tasks.  The net result of
+ * all this is that, if stable variables are multiply assigned, the
+ * program execution may observe any "historical" value (if it was
+ * captured by some particular compilation task), as well as a "most
+ * recent" value observed by the interpreter or less-optimized code.
+ * <p>
+ * For all these reasons, a user who bends the rules for a stable
```

Comment:
I've cases in the JDK where a stable fields could, under race, be assigned to objects that are semantically equivalent, but whose identity is distinct. Would you say this is a good use of `@Stable` ? In FFM we took extra precaution and deduplicated the identity by using a concurrent hash map -- but is this required?

Reply:
I think so; we have benign races for non-stable fields like `Class.reflectionFactory`. However, sometimes sharing the same object avoids redundant cache computation. It is ultimately a tradeoff between synchronization and lock overheads and the applicability of cache.